bundles/automation/org.eclipse.smarthome.automation.providers/src/main/java/org/eclipse/smarthome/automation/core/provider/AutomationResourceBundlesEventQueue.java

```diff
@@ -0,0 +1,312 @@
+/**
+ * Copyright (c) 1997, 2015 by ProSyst Software GmbH and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.automation.core.provider;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.eclipse.smarthome.automation.Rule;
+import org.eclipse.smarthome.automation.template.TemplateProvider;
+import org.eclipse.smarthome.automation.type.ModuleTypeProvider;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleEvent;
+import org.osgi.util.tracker.BundleTracker;
+import org.osgi.util.tracker.BundleTrackerCustomizer;
+
+/**
+ * This class is responsible for tracking the bundles providing automation resources and delegating the processing to
+ * the responsible providers in separate thread.
+ *
+ * @author Ana Dimova - Initial Contribution
+ * @author Kai Kreuzer - refactored (managed) provider and registry implementation
+ *
+ */
+class AutomationResourceBundlesEventQueue implements Runnable, BundleTrackerCustomizer<Object> {
+
+    /**
+     * This static field serves as criteria for recognizing bundles providing automation resources. If these bundles
+     * have such manifest header this means that they are such providers.
+     */
+    public static final String AUTOMATION_RESOURCES_HEADER = "Automation-ResourceType";
```

Comment:
Question about this header: Do you really think that we need it? From our analysis, it should be equally fast to check the existence of the ESH-INF/automation resource to check whether a bundle has automation resources or not. Requiring a manifest header is a possible reason for problems, since people might tend to forget adding it.